File: src/algorithms/pathfinding/graph-traversal/iterative-deepening-dfs/sources/iterative-deepening-dfs_test.py
import importlib

iterative_deepening_dfs_mod = importlib.import_module("iterative-deepening-dfs")
iterative_deepening_dfs = iterative_deepening_dfs_mod.iterative_deepening_dfs


def make_empty_grid(rows, cols):
    return [[{"type": "empty"} for _ in range(cols)] for _ in range(rows)]


def set_cell(grid, row, col, cell_type):
    grid[row][col]["type"] = cell_type


def test_finds_path_on_empty_grid():
    grid = make_empty_grid(4, 4)
    result = iterative_deepening_dfs(grid, (0, 0), (3, 3))
    assert len(result["path"]) > 0
    assert result["path"][0] == (0, 0)
    assert result["path"][-1] == (3, 3)


def test_finds_shortest_path():
    grid = make_empty_grid(1, 5)
    result = iterative_deepening_dfs(grid, (0, 0), (0, 4))
    assert len(result["path"]) == 5


def test_returns_empty_path_when_no_route():
    grid = make_empty_grid(3, 3)
    set_cell(grid, 0, 1, "wall")
    set_cell(grid, 1, 0, "wall")
    set_cell(grid, 1, 1, "wall")
    result = iterative_deepening_dfs(grid, (0, 0), (2, 2))
    assert result["path"] == []


def test_handles_adjacent_start_and_end():
    grid = make_empty_grid(3, 3)
    result = iterative_deepening_dfs(grid, (0, 0), (0, 1))
    assert len(result["path"]) == 2
    assert result["path"][0] == (0, 0)
    assert result["path"][1] == (0, 1)


def test_depth_reached():
    grid = make_empty_grid(1, 4)
    result = iterative_deepening_dfs(grid, (0, 0), (0, 3))
    assert result["depthReached"] == 3


def test_path_is_valid_adjacent_steps():
    grid = make_empty_grid(4, 4)
    result = iterative_deepening_dfs(grid, (0, 0), (3, 3))
    for path_index in range(1, len(result["path"])):
        prev = result["path"][path_index - 1]
        curr = result["path"][path_index]
        assert abs(curr[0] - prev[0]) + abs(curr[1] - prev[1]) == 1


def test_tracks_visited_cells():
    grid = make_empty_grid(3, 3)
    result = iterative_deepening_dfs(grid, (0, 0), (2, 2))
    assert len(result["visited"]) > 0


if __name__ == "__main__":
    test_finds_path_on_empty_grid()
    test_finds_shortest_path()
    test_returns_empty_path_when_no_route()
    test_handles_adjacent_start_and_end()
    test_depth_reached()
    test_path_is_valid_adjacent_steps()
    test_tracks_visited_cells()
    print("All tests passed!")

File: src/algorithms/pathfinding/graph-traversal/iterative-deepening-dfs/__tests__/iterative-deepening-dfs_test.go
package iterativedeepeningdfs

import "testing"

func makeEmptyGrid(rows, cols int) [][]GridCell {
	grid := make([][]GridCell, rows)
	for row := range grid {
		grid[row] = make([]GridCell, cols)
		for col := range grid[row] {
			grid[row][col] = GridCell{Row: row, Col: col, CellType: CellEmpty, State: "default"}
		}
	}
	return grid
}

func setWallCell(grid [][]GridCell, row, col int) {
	grid[row][col].CellType = CellWall
}

func TestFindsPathOnEmptyGrid(t *testing.T) {
	grid := makeEmptyGrid(4, 4)
	result := IterativeDeepeningDfs(grid, 0, 0, 3, 3)
	if len(result.Path) == 0 {
		t.Error("expected non-empty path")
	}
	if result.Path[0][0] != 0 || result.Path[0][1] != 0 {
		t.Errorf("expected path start [0,0]")
	}
	last := result.Path[len(result.Path)-1]
	if last[0] != 3 || last[1] != 3 {
		t.Errorf("expected path end [3,3]")
	}
}

func TestFindsShortestPath(t *testing.T) {
	grid := makeEmptyGrid(1, 5)
	result := IterativeDeepeningDfs(grid, 0, 0, 0, 4)
	if len(result.Path) != 5 {
		t.Errorf("expected path length 5, got %d", len(result.Path))
	}
}

func TestReturnsEmptyPathWhenNoRoute(t *testing.T) {
	grid := makeEmptyGrid(3, 3)
	setWallCell(grid, 0, 1)
	setWallCell(grid, 1, 0)
	setWallCell(grid, 1, 1)
	result := IterativeDeepeningDfs(grid, 0, 0, 2, 2)
	if len(result.Path) != 0 {
		t.Errorf("expected empty path, got %d steps", len(result.Path))
	}
}

func TestHandlesAdjacentStartAndEnd(t *testing.T) {
	grid := makeEmptyGrid(3, 3)
	result := IterativeDeepeningDfs(grid, 0, 0, 0, 1)
	if len(result.Path) != 2 {
		t.Errorf("expected path length 2, got %d", len(result.Path))
	}
}

func TestDepthReached(t *testing.T) {
	grid := makeEmptyGrid(1, 4)
	result := IterativeDeepeningDfs(grid, 0, 0, 0, 3)
	if result.DepthReached != 3 {
		t.Errorf("expected depthReached 3, got %d", result.DepthReached)
	}
}
